docs/design/2022-04-21-performance-diagnosis-enhancements.md

```diff
@@ -0,0 +1,188 @@
+- Author(s): [cfzjywxk](https://github.com/cfzjywxk), [you06](https://github.com/you06)
+- Discussion PR:
+- Tracking Issue: #33963, #34106
+
+## Table of Contents
+
+* [Introduction](#introduction)
+* [Motivation or Background](#motivation-or-background)
+* [Detailed Design](#detailed-design)
+* [Impacts & Risks](#impacts--risks)
+
+## Introduction
+
+To enhance the diagnosis and tuning of TiDB, this RPC proposes some optimization of metrics, including the supplement of metrics, better display in Grafana.
+
+## Motivation or Background
+
+TiDB collects many metrics, but few of them are frequently used and the metrics of the critical paths are in different panels, which makes the diagnosis a problem. Furthermore, some latency events are not recorded in critical paths.
+
+This RPC proposes some improvements to the metrics collection, and display.
+
+## Detailed Design
+
+### Differences Between Internal and External Requests
+
+The tracking issue of it is #33963, and it should be applied to most of the metrics.
+
+#### Internal Transaction Usage
+
+The main goal of this task is to differentiate the internal and external transactions and mark the snapshots and transactions as internal for internal sessions and tasks.
+
+Some tasks are not running in the internal sessions, however, they are executed directly by `RunInNewTxn` :
+
+- DDL Backfill
+- Auto-ID management
+- Version check when bootstrapping
+- GC placement and label rules
+- Load data for cached tables
+- Alloc global ID for some operations
+    - Write sequence update to binlog
+    - Create temporary table
+
+When TiDB bootstrapping, it’ll create some internal sessions, these sessions are used for:
+
+- Bind info read/write
+- Privilege read/write
+- System variable read/write
+- Telemetry report
+- Statistic load
+
+The coprocessor request, transaction, and snapshot created by `RunInNewTxn` and the internal sessions need to be marked as internal.
+
+#### kvproto
+
+Add a field `request_source` in the proto, which is a string type field and can be empty(TiDB does not allow empty `request_source`). The string type can make the metrics display more readable.
+
+```diff
+// Miscellaneous metadata attached to most requests.
+message Context {
+    reserved 4;
+    ...
++   // The source of the request.
++   string request_source = xx;
```

Comment:
How about to encode such info into [task_id](https://github.com/pingcap/kvproto/blob/f7a7c8ccda74ae17acdde7b47cf829ac6f5c2a52/proto/kvrpcpb.proto#L695)? eg. the first bit indecates the scope.

Reply:
> It may be defined as a string tag, or a simple integer tag with its real meaning defined in TiDB.

From @sticnarf's comment, kvproto should not be coupled with TiDB since the tag may also be used by other clients.

Encode it in the higher bits of `task_id`, 1 bit for internal and external, and at least 5 extra bits for source types of txn, maybe 8 bits should be reserved. But it's too tricky to encode this into `task_id` (

Reply:
Agreed. It's efficient but little tricky. Currently, the task id is just a instance-level in-memory atomic integer, I guess it's not strongly coupled with tidb, for example, if some one use client-go directly and not set context.task_id directly, it's always 0. The problem is, tikv needs to known the encode method for reporting human-readable metrics.